### PR TITLE
Feat/206 account balance staking support

### LINKED ISSUE #206: `/account/balance` staking support
As we will be introducing staking functionality, we also need to be able to return staking account balances. We need to provide the balance of a staking account at a certain block number. To do so, we need to let people to query for an account based on a stake address.

The response format should not change but different tables will need to be queried when looking for a stake address.

1. [This query](https://github.com/input-output-hk/cardano-db-sync/blob/master/doc/interesting-queries.md#get-the-reward-history-for-a-specified-stake-address) could be used as as a starting point although don't forget to consider the withdrawals.
2. We can switch thery by address `bech32` preffix (`stake`, `stake_test`)
3. Update `src/server/services/block-service.ts#findUtxoByAddressAndBlock` to:
   1. Switch based on the address type and perform different queries
   2. Update the interface to return something for stake and non stake accounts.
   3. Update `src/server/utils/data-mapper.ts#mapToAccountBalanceResponse` to properly map the new type (do not include `utxos` in case of staking

## Testing

As mentioned in #208 we cannot use a db-snapshot therefore some blocks will need to be inserter to fake data.

1. Check our e2e suite is still working (no actions needed)
2. Add new tests to the current ones. To do so, we will need to _manually_ insert a fake blocks:
     - A block were a reward is received
     - A block were another reward is received (in order to test the sum is working) 
     - A block with a withdrawal to check staking account balance has been reduced
3. We need to be make sure `rosetta-cli check:data` still works
   1. Deploy the new version (using the commit as build context for docker) on our mainnet testing server.
   2. Execute `check:data` for a block range containing block `TBD`.
   3. Execute `check:data` for the whole chain.

### PR DESCRIPTION
**Pull request by jvieiro**
# Description

Endpoint /account/balance now supports stake accounts.

# References

Closes #206.

### CODE REVIEW

File: cardano-rosetta-server/src/server/db/blockchain-repository.ts [hunk 1]
```diff
@@ -63,6 +64,13 @@ export interface BlockchainRepository {
    * @param blockIdentifier block information, when value is not undefined balance should be count till requested block
    */
   findUtxoByAddressAndBlock(logger: Logger, address: string, blockHash: string): Promise<Utxo[]>;
+
+  /**
+   * Returns an array containing all utxo for address till block identified by blockIdentifier if present, else the last
```
Comment (AlanVerbner) [inline cardano-rosetta-server/src/server/db/blockchain-repository.ts:69]: careful as this are not utxos

File: cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts [hunk 2]
```diff
@@ -132,7 +132,7 @@ LIMIT 1
 `;
 
 export interface FindBalance {
-  balance: string;
+  balance: number;
```
Comment (AlanVerbner) [inline cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts:135]: It needs to be a string as it's a `BitInteger` not a `number`

File: cardano-rosetta-server/src/server/models.ts [hunk 3]
```diff
@@ -60,3 +60,8 @@ export interface BlockUtxos {
   block: Block;
   utxos: Utxo[];
 }
+
+export interface BlockBalance {
+  block: Block;
+  balance: number;
```
Comment (AlanVerbner) [inline cardano-rosetta-server/src/server/models.ts:66]: Same as before

File: cardano-rosetta-server/src/server/models.ts [hunk 4]
```diff
@@ -60,3 +60,8 @@ export interface BlockUtxos {
   block: Block;
   utxos: Utxo[];
 }
+
+export interface BlockBalance {
```
Comment (AlanVerbner) [inline cardano-rosetta-server/src/server/models.ts:64]: I'm not sure this name is appropriate. To me it seems it refers to the balance a block has (which I don't think makes sense). Maybe `BalanceAtBlock` or something like that makes more sense.

File: cardano-rosetta-server/src/server/utils/constants.ts [hunk 5]
```diff
@@ -11,6 +11,8 @@ export const VIN = 'Vin';
 export const VOUT = 'Vout';
 export const SIGNATURE_TYPE = 'ed25519';
 
+export const stakeTypes = ['stake', 'stake_test'];
```
Comment (AlanVerbner) [inline cardano-rosetta-server/src/server/utils/constants.ts:14]: I think this string sa re used on another file and maybe they should be constants / enums

File: cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts [hunk 6]
```diff
@@ -166,6 +165,24 @@ const selectUtxoDetail = `SELECT
...
 
+const findBalanceByAddressAndBlock = `SELECT (SELECT COALESCE(SUM(r.amount),0) 
+  FROM reward r
+  JOIN stake_address ON 
+    stake_address.id = r.addr_id AND 
+    stake_address.view = $1
+  JOIN block ON
+    block.id = r.block_id
+  WHERE stake_address.view = $1
+  AND block.epoch_no < (SELECT epoch_no FROM block WHERE hash = $2))-
```
Comment (AlanVerbner) [inline cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts:176]: Why not `<=` ?

File: cardano-rosetta-server/src/server/services/block-service.ts [hunk 7]
```diff
@@ -133,17 +147,31 @@ const configure = (repository: BlockchainRepository): BlockService => ({
...
     if (block === null) {
-      logger.error('[findUtxoByAddressAndBlock] Block not found');
+      logger.error('[findBalanceDataByAddressAndBlock] Block not found');
       throw ErrorFactory.blockNotFoundError();
     }
-    logger.info(`[findUtxoByAddressAndBlock] Looking for utxos for address ${address} and block ${block.hash}`);
+
+    logger.info(`[findBalanceDataByAddressAndBlock] Looking for utxos for address ${address} and block ${block.hash}`);
+    const addressPrefix = cardanoService.getPrefixFromAddress(address);
+    if ([stakeType.STAKE as string, stakeType.STAKE_TEST as string].some(type => addressPrefix.includes(type))) {
```
Comment (AlanVerbner) [inline cardano-rosetta-server/src/server/services/block-service.ts:159]: can we extract this to a function?

File: cardano-rosetta-server/test/e2e/account/account-api.test.ts [hunk 8]
```diff
@@ -287,4 +287,33 @@ describe('/account/balance endpoint', () => {
       coins: []
     });
   });
+  test('should return 0 for the balance of stake account at block with no earned rewards', async () => {
```
Comment (AlanVerbner) [inline cardano-rosetta-server/test/e2e/account/account-api.test.ts:290]: @jvieiro aren't we missing some tests here?

File: cardano-rosetta-server/src/server/services/block-service.ts (hunk 7 above)
Comment (jvieiro) [reply cardano-rosetta-server/src/server/services/block-service.ts]: Done at https://github.com/input-output-hk/cardano-rosetta/pull/213/files#diff-3f8985087fb8b875b0466a3ece1b03ad5f0201cc711b9b087992370e949b30e9R357.

File: cardano-rosetta-server/test/e2e/account/account-api.test.ts (hunk 8 above)
Comment (jvieiro) [reply cardano-rosetta-server/test/e2e/account/account-api.test.ts:290]: Yes, https://github.com/input-output-hk/cardano-rosetta/pull/213/files#diff-3f3a229c474782e5b82a3cc97895e2d051e3cbe44b9452eb31bcdc073dea053eR319 checks rewards sum and withdrawals subtraction

File: cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts (hunk 6 above)
Comment (jvieiro) [reply cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts:175]: I thought it was a requirement that the `epoch_no` where the reward was earned has to be less than the `epoch_no` of the block that was consulted.

Comment (AlanVerbner) [reply cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts]: I was just asking, I'm not sure either

File: cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts [hunk 9]
```diff
@@ -166,6 +165,23 @@ const selectUtxoDetail = `SELECT
...
 const findUtxoByAddressAndBlock = findUtxoFieldsByAddressAndBlock(selectUtxoDetail);
 
+const findBalanceByAddressAndBlock = `SELECT (SELECT COALESCE(SUM(r.amount),0) 
+  FROM reward r
+  JOIN stake_address ON 
+    stake_address.id = r.addr_id
+  JOIN block ON
+    block.id = r.block_id
+  WHERE stake_address.view = $1
+  AND block.epoch_no <= (SELECT epoch_no FROM block WHERE hash = $2))-
```
Comment (AlanVerbner) [inline cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts:175]: I just took a second look at the queries. The `reward.block_id` links to the block where the the reward was actually paid so ... why not to do `block.id <= (select id from block where hash = $2)` ? I'm not quie sure if I'm (now) convinced by the epoch filter

File: cardano-rosetta-server/test/e2e/construction/construction-metadata-api.test.ts [hunk 10]
```diff
@@ -47,7 +47,7 @@ describe(CONSTRUCTION_METADATA_ENDPOINT, () => {
     expect(response.statusCode).toEqual(StatusCodes.OK);
     expect(response.json()).toEqual({
       metadata: {
-        ttl: (latestBlock.block.metadata.slotNo + relativeTtl).toString()
+        ttl: (latestBlockSlot + relativeTtl).toString()
```
Comment (AlanVerbner) [inline cardano-rosetta-server/test/e2e/construction/construction-metadata-api.test.ts:50]: is this change related to your PR?

File: cardano-rosetta-server/test/e2e/account/account-api.test.ts [hunk 11]
```diff
@@ -287,4 +287,62 @@ describe('/account/balance endpoint', () => {
...
+          value: '0',
+          currency: {
+            decimals: 6,
+            symbol: 'ADA'
+          }
+        }
+      ]
+    });
+  });
+  test('should sum all rewards and subtract all withdrawals till block 4876885', async () => {
```
Comment (AlanVerbner) [inline cardano-rosetta-server/test/e2e/account/account-api.test.ts:319]: why have you picked this block? It would help me undersand the values

File: cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts (hunk 9 above)
Comment (jvieiro) [reply cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts:175]: I was thinking of filter by epoch to check if the rewards were paid at some point.  I used `<`operator at the first attempt because if the `epoch_no` equals  `block.epoch_no` it means that the rewards haven't been paid because the epoch's period hasn't finished at this point, right? If I am wrong and  `<=` operator is OK I think we can use `block.id` as a filter.

File: cardano-rosetta-server/test/e2e/account/account-api.test.ts (hunk 11 above)
Comment (jvieiro) [reply cardano-rosetta-server/test/e2e/account/account-api.test.ts:319]: Because at this point the following rewards and withdrawals were introduced:
- Rewards: `112588803` at block `4597956` and  `111979582` at block `4619398`
- Withdrawals: `112588803` at block `4598038`

File: cardano-rosetta-server/test/e2e/construction/construction-metadata-api.test.ts (hunk 10 above)
Comment (jvieiro) [reply cardano-rosetta-server/test/e2e/construction/construction-metadata-api.test.ts]: Not really, I've added this to keep the tests working, the change is the same done in #208's PR.

File: cardano-rosetta-server/test/e2e/account/account-api.test.ts (hunk 11 above)
Comment (AlanVerbner) [reply cardano-rosetta-server/test/e2e/account/account-api.test.ts:319]: can you leave a comment about this pls? because if someone new approaches this test won't understand

File: cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts (hunk 9 above)
Comment (AlanVerbner) [reply cardano-rosetta-server/src/server/db/queries/blockchain-queries.ts:175]: Let's do this:

1. Change it to use the `block.id`
2. Check if your tests and queries and everything still works

On a separate thread I will try to confirm what i'm saying

**Review by tomasBustamante — APPROVED** [review]
